File: Data Structures & Algorithms/balanced-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    balanced = true;

    isBalanced(root) {
        if (root === null) return true;
        this.dfs(root);
        return this.balanced;
    }

    dfs(node) {
        if (node === null) return 0;
        var left = this.dfs(node.left);
        var right = this.dfs(node.right);
        if(Math.abs(left - right) > 1) this.balanced = false;
        return Math.max(left,right) + 1;
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    ArrayDeque<Integer> history = new ArrayDeque<>();
    public int kthSmallest(TreeNode root, int k) {
        inorder(root);
        int kthSmallest = -1;
        for (int i = 0; i < k; i++) {
            kthSmallest = history.removeFirst();
        }
        return kthSmallest;
    }

    private void inorder(TreeNode root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        history.addLast(root.val);
        inorder(root.right);
    }
}
